RecoHGCal/TICL/plugins/LinkingAlgoBase.h

```diff
@@ -0,0 +1,52 @@
+#ifndef RecoHGCal_TICL_LinkingAlgoBase_H__
+#define RecoHGCal_TICL_LinkingAlgoBase_H__
+
+#include <vector>
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "DataFormats/HGCalReco/interface/Trackster.h"
+#include "DataFormats/HGCalReco/interface/TICLCandidate.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/MuonReco/interface/Muon.h"
+#include "MagneticField/Engine/interface/MagneticField.h"
+#include "TrackingTools/GeomPropagators/interface/Propagator.h"
+#include "RecoLocalCalo/HGCalRecAlgos/interface/RecHitTools.h"
+#include "Geometry/HGCalCommonData/interface/HGCalDDDConstants.h"
+
+namespace edm {
+  class Event;
+  class EventSetup;
+}  // namespace edm
+
+namespace ticl {
+  class LinkingAlgoBase {
+  public:
+    LinkingAlgoBase(const edm::ParameterSet& conf) : algo_verbosity_(conf.getParameter<int>("algo_verbosity")) {}
+
+    virtual ~LinkingAlgoBase(){};
+
+    virtual void initialize(const HGCalDDDConstants* hgcons,
+                            const hgcal::RecHitTools rhtools,
+                            const edm::ESHandle<MagneticField> bfieldH,
+                            const edm::ESHandle<Propagator> propH) = 0;
+
+    virtual void linkTracksters(const edm::Handle<std::vector<reco::Track>> tkH,
+                                const edm::ValueMap<float>& tkTime,
+                                const edm::ValueMap<float>& tkTimeErr,
+                                const edm::ValueMap<float>& tkTimeQual,
+                                const std::vector<reco::Muon>& muons,
+                                const edm::Handle<std::vector<Trackster>> tsH,
+                                std::vector<TICLCandidate>& resultTracksters) = 0;
+
+    static void fillPSetDescription(edm::ParameterSetDescription& desc) { desc.add<int>("algo_verbosity", 0); };
+
+    enum VerbosityLevel { None = 0, Basic, Advanced, Expert, Guru };
```

Comment:
I would rather prefer not to introduce a module-specific logging configuration and just stick with pure MessageLogger levels.

If you keep this, not only you toyda, but every future reader and maintainer will have two separate levels to keep in mind and get right, rather than one.

Reply:
Hi @jpata , thanks for the comment. Eventually it would be nice to have it uniform across CMSSW with Message Logger levels. 
I asked the active developers if this existing code is something they are actively using in their debugging workflows, and this module is still under heavy development. Do you mind if we keep it like this until the development of this module becomes maintenance and then we roll back not only linking but also other modules that are using these message levels?

Reply:
Anything that is merged is already maintenance and someone other than the immediate developers may have to step in to work on it.
I doubt consistent logging will be implemented in the future, if it's already too troublesome to do now.

If you decide to keep this, you could consider using the same enum across different TICL modules, so at least it's possible to keep self consistent in TICL in the future?
Now I see redefinitions of this VerbosityLevel in PatternRecognitionAlgoBase, HGCGraph, SeedingRegionAlgoBase.
For modules in this PR that take a bool flag `debug`, you can prefer the enum, so all of them can be configured in the same way.

Reply:
I have replaced the class wise enums with something that is common for all the TICL modules where this is used